Background image: move controls into a popover

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR:

1. moves the background controls into a flyout panel
2. consolidates properties in the tool panel beneath the "background image" property, which means that resetting an image will reset all associated properties

See: 
- https://github.com/WordPress/gutenberg/pull/57005#issuecomment-1866076523
- https://github.com/WordPress/gutenberg/pull/60100#issuecomment-2178866354

Doing this with a view to consolidate colors and fill. See: https://github.com/WordPress/gutenberg/pull/60100#issuecomment-2160054475

Also take into account:

- https://github.com/WordPress/gutenberg/pull/59454#issuecomment-2020150910
- https://github.com/WordPress/gutenberg/pull/59454#issuecomment-2021680599



### TODO

- [x] Refactor main components in packages/block-editor/src/components/global-styles/background-panel.js
- [x] Condense and consolidate controls in the popover
- [x] Mobile view - ensure it works as expected. 
- [x] Test keyboard interaction
- [ ] ~Wire in tools panel reset for individual properties? E.g., position and repeat?~

## Why?

To win back space in the sidebar controls.

## How?

Moving background image controls into a flyout popover. This includes, where an image exists, the media select control.

To get the Tools Panel controls to work, I've created an invisible `ToolsPanelItem` components for each control set - this registers reset callbacks and so on in the Tools Panel.

## Testing Instructions

The flyout appears in the Global styles (Layout > Background Image) and at the block level for Group, Quote, Post Content blocks. For both:

1. Check that you add/upload/reset background images from the sidebar panel.
3. When a background image is selected, the sidebar control activates the flyout.
4. The controls the flyout should update the background image properties as expected
5. Ensure the Tools Panel controls works, that is, everything should open, close and reset.
6. In mobile view (< 782px) check that the flyout appears in the viewport width. 
7. Resetting the image should also reset size, position, and repeat.

### Testing Instructions for Keyboard

Check:

1. You can tab through the sidebar controls, and add a background image
2. . Where a background image exists, check you can tab to the background image control and hit Enter to open and focus the new flyout.
3. . Tab through the flyout and change the background image properties


## Screenshots or screencast <!-- if applicable -->

| Global styles  | Blocks |
| ------------- | ------------- |
| <img width="300" alt="Screenshot 2024-06-27 at 11 47 29 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/ab08002d-e7a4-419e-afb4-3f8cc59481af">  | <img width="300" alt="Screenshot 2024-06-27 at 11 48 01 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/b115bb43-d7a3-48af-9d30-e5f8c61c9ea8"> |

Resetting background properties:


https://github.com/WordPress/gutenberg/assets/6458278/13137e1a-893e-44a4-bc1d-af7d9a9742bf



## Related

- https://core.trac.wordpress.org/ticket/22058

ramonjd committed:
Testing other component layout and dummy toolspanel items
Refactor the media toggle button and refactor controls
Temp. don't send default controls in hook

ramonjd commented:
Notes:

There's a bit of a blocker on this particular design since I'm not sure I can integrate tools panel functionality (show by default / reset) with the popover contents. 

The tools panel registers rendered tools panel items. Given that the background controls are in the dropdown popover, the items aren't rendered and so the tools panel ellipsis menus don't appear.

I've tried adding a "dummy" `ToolsPanelItem` in the DOM. It kinda works, but it's not consistent with other block supports and style options. Furthermore, it breaks the block supports `__experimentalDefaultControls`.

ramonjd committed:
Rejigging components, catering for no-background-image-controls

ramonjd committed:
With tabs

A bit of margin

ramonjd commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -502,93 +560,97 @@ function BackgroundSizeToolsPanelItem( {
+				focusable={ false }
+				className="block-editor-global-styles-background-panel__tab-panel"
+			>
+				<FocalPointPicker
```
One thing to note about the focal point picker is that its dimensions vary - it can be very tall or short depending on the image. This is whether it's in a popover or not, but it affect the popover height and scroll.

A tall image:

https://github.com/WordPress/gutenberg/assets/6458278/dd4162e9-f963-4a60-a733-934c29dfcdf4

Constraining it via max-height doesn't make the tool any more usable:

<img width="593" alt="Screenshot 2024-06-20 at 3 30 20 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/e06cf7a5-e841-43cf-aa26-7184a6b6d8d8">

jasmussen commented:
Nice work. Took a stab at some designs in [this Figma](https://www.figma.com/design/hrrqkZCCRmbO8A2VX0aKun/Background-Image?node-id=15-2813&t=8WYZ41SggSPEWbsX-1). The main thing I wanted to address was to illustrate the idea of just moving the entire background image panel inside the flyout. That might also address the split behavior of this one, which doesn't feel as ergonomic as it could:
![status](https://github.com/WordPress/gutenberg/assets/1204802/975bf698-a184-4a2b-8c60-ab54ebf2a3ac)

Here are the results, but note that they should be thought of as _in a draft state_. @jameskoster shared some ideas that went in a different direction on another of these issues, and I'd like him to have a chance to chime in, and perhaps remix the figma pieces.

Group block with no background image:

<img width="2800" alt="Group block with no background image" src="https://github.com/WordPress/gutenberg/assets/1204802/0eaad9d5-d260-4303-901b-aa7f706557e8">

Note how even the "Add background image" button was moved inside the flyout.

Group block _with_ a background image:

<img width="2800" alt="Group block with background image" src="https://github.com/WordPress/gutenberg/assets/1204802/a6aecf32-632f-4002-88ed-b54dc3459f3f">

The flyout essentially contain the contents of the panel, including tiling and position options. I didn't see a clear need for tabs yet, but as noted, Jay had some thoughts too that I wanted to give him a chance to illustrate.

Additional note, this mockup tries a different design for the background image "swatch"—instead of being a circle, it is here, a square, to better match what we are showing in the List View for images, etc.

jameskoster commented:
> Note how even the "Add background image" button was moved inside the flyout.

This feels a bit of a drawback since it's adding a click. When there's no image set, could the button open the media library directly? There are options to upload or select an image there.

> Group block with a background image

This mockup seems like a good direction for now. We're just missing controls to replace or unset the background image.

For unsetting there are a couple of existing patterns to consider:

| Clearing a color | Removing a shadow |
| --- | --- |
| <img width="260" alt="Screenshot 2024-06-20 at 11 49 13" src="https://github.com/WordPress/gutenberg/assets/846565/1dbaadd8-93c5-458e-9de1-4a901a9729cf"> | <img width="280" alt="Screenshot 2024-06-20 at 11 54 17" src="https://github.com/WordPress/gutenberg/assets/846565/d0b8a124-ef1e-4908-97e6-25e3711d5598"> |

Having a 'Clear' button in the popover might be strange to use, because clicking it would hide all the other settings. For that reason I prefer the latter. It also makes unsetting require one less click.

Replacing is slightly tricker. It's probably okay to place this affordance in the popover, perhaps as an `Item` which opens the Media Library on click?

<img width="344" alt="Screenshot 2024-06-20 at 12 48 31" src="https://github.com/WordPress/gutenberg/assets/846565/1ae0cc0a-4ad2-48f9-9cde-f8306d915e82">

jasmussen commented:
The minus to unset is a great argument for making this follow the ItemGroup pattern. And feel free to see if we can keep the "Add background image" CTA right in the Item itself. But we need some way to then "replace" from _inside_ the flyout, it probably shouldn't be that recycle button.

ramonjd commented:
Thanks for the continued support on this @jasmussen and @jameskoster 

Also for the Figma link, it helps a lot!

> > Note how even the "Add background image" button was moved inside the flyout.
> 
> This feels a bit of a drawback since it's adding a click. When there's no image set, could the button open the media library directly? There are options to upload or select an image there.

This is very doable. So, 

1. Where no image is set, the button has the select/upload dropdown menu
2. Where an image is set, the preview icon is shown on the button, and the flyout appears. The media library controls are now in the flyout as per the design above?

I'll try it out in the next iteration. 

As mentioned over at https://github.com/WordPress/gutenberg/pull/60100#issuecomment-2179586732 it'd be nice to get the flyout/refactor done first and then work on relocating the background controls, e.g., in Color and Fill. These PRs get very chunky very quickly 😄

jasmussen commented:
Took a stab at an i3 that incorporates the conversation of the reset button being added, and the initial state simply opening the media library. It's not quite working so well:

<img width="2800" alt="Background image tools" src="https://github.com/WordPress/gutenberg/assets/1204802/27e36a24-b7a6-48e1-976d-7b8312c84d7b">

I still think these are the right ingredients, with the majority of work happening in the modal. But there's something here that isn't connecting, and it's all related to @jameskoster's excellent [observations here](https://github.com/WordPress/gutenberg/pull/60100#issuecomment-2180553281). 

The short is, I still think we probably want this panel to live in a flyout. But probably it's good to establish first: _where does the button live?_ Jay, let me know if this inspires any ideas.

jameskoster commented:
In the short/medium term I don't mind putting the button in the Color (& Fill) panel. I also wouldn't mind leaving the Background Image panel and placing it there. I think both provide a pathway to a dedicated 'Background' or 'Fill' panel in the future, once we've fully explored that idea.

ramonjd committed:
Revert tabs and move the image upload to the floating panel

ramonjd commented:
Testing out moving the media replace flow to the flyout panel:


https://github.com/WordPress/gutenberg/assets/6458278/e3ec64d2-22d5-4447-bbe0-07ed79e8aa23

A couple of janky issues, mainly the popover has a z-index of `1000000` 😱 so it sits on top of the media library:


<img width="1149" alt="Screenshot 2024-06-25 at 12 43 22 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/865a7a22-3bcb-4d2b-ae8a-839fab12c331">

My instinct tells me the media library should always sit on top of everything, so bumping its z-index could work. The tricker path is closing the background image controls panel when the media library is open.

andrewserong commented:
Looks promising, though!

> My instinct tells me the media library should always sit on top of everything, so bumping its z-index could work.

It seems we're doing that for a bunch of other modals, so sounds like it could be a good approach?

https://github.com/WordPress/gutenberg/blob/044a05f1486f2631675ac3a98ddd64018c63e574/packages/base-styles/_z-index.scss#L124-L135

ramonjd committed:
Temp hack to push popover into the background

ramonjd committed:
Fix hack to push popover into the background

ramonjd commented:
> It seems we're doing that for a bunch of other modals, so sounds like it could be a good approach?

That's the ticket. Found a media library-related one too. Thanks!

jameskoster commented:
This seems to be working quite well! A couple of small details to fix:

<img width="91" alt="Screenshot 2024-06-25 at 09 48 39" src="https://github.com/WordPress/gutenberg/assets/846565/e91175c5-04b6-45ec-a39b-402d45c52de0">

The gap between the swatch and the label should be the same as the color button.

---

<img width="283" alt="Screenshot 2024-06-25 at 10 36 33" src="https://github.com/WordPress/gutenberg/assets/846565/d16c0e85-0a50-4efd-af09-faa693e10757">

Could this menu be the same width as the button?

ramonjd committed:
Tweak size of media flow dropdown
Icon margin tweak

ramonjd commented:
> The gap between the swatch and the label should be the same as the color button.
> Could this menu be the same width as the button?

Thanks for catching that. This should be done 👍🏻 

One question I had, which can be for another PR: should resetting the image also reset other background image settings?

So, if I remove an image, should we also wipe size, position etc?

https://github.com/WordPress/gutenberg/assets/6458278/f2cebae9-7b53-481f-8f98-bc6c349597cf

I'm thinking that if an image is wiped, it'd be okay to remove associated properties given the variation between images. 🤷🏻

jasmussen commented:
> One question I had, which can be for another PR: should resetting the image also reset other background image settings?

When it's part of the toolspanel, no, mainly because as far as possible, we want to avoid exceptions and edgecases in the behavior. That said, now that those controls are moving into the flyout, they should no longer be considered separate controls in the toolspanel at all, they stop being individual design tools, and become part of the singular "Background image" tool that exists in the flyout. 

However before you go ahead and change this behavior, I'd love to loop in @jameskoster, per recent conversations around how unifying background tools (https://github.com/WordPress/gutenberg/issues/47369#issuecomment-2191737305). If we want individual size and position tools for each stacked background layer, it's probably fine to go ahead and remove these from the toolspanel, and allow them to be connected to each background layer. If we want something that applies across, it needs separate thinking.

Pretty sure I saw a separate conversation that suggested there be individual properties per background, so that would help decide the above.

jameskoster commented:
Yes, it will be important for each layer in the background stack to have its own settings. So I agree "Size" should be removed from toolspanel for now. 

To answer the original question; resetting the background image should also reset the associated size/position values.

brentjett commented:
Regarding the media modal and popovers showing on top - have you guys considered swapping in a `<dialog>` element so the media picker would display in the top-level? Dialogs are well supported now and can make the rest of the page inert. They aren't subject to z-index issues because the top-layer is a latest-takes-precedence context.

ramonjd commented:
> Regarding the media modal and popovers showing on top - have you guys considered swapping in a <dialog> element so the media picker would display in the top-level? 

Thanks for the comment. 

I don't think it's come up, but it looks like [<dialog>.showModal()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal) has pretty good support. 

If work on the modal is indeed part of it, I think should be a consideration for ongoing [Media Library tasks](https://github.com/WordPress/gutenberg/issues/55238). 

Also potentially related:

- https://github.com/WordPress/gutenberg/issues/55894

ramonjd commented:
> they stop being individual design tools, and become part of the singular "Background image" tool that exists in the flyout.
> Yes, it will be important for each layer in the background stack to have its own settings. So I agree "Size" should be removed from toolspanel for now.

Thanks again for confirming, folks. 

This works out well as I can delete my dirty hack that injects size into the Tools Panel and no-one has to know 🙈

ramonjd committed:
Remove default controls override as there's only 'background image' now.
Reset image functions reset all background image properties.

isShownByDefault prop not used. Deleting

Prevent focal point picker creating horizontal scroll

Expand popover on mobile
Don't resize with window height change

ramonjd commented:
All updated. I think this PR is close. Once it lands I'll carry on with:

- https://github.com/WordPress/gutenberg/pull/60100

andrewserong reviewed (commented):
Great work here, it's generally working well for me 👍

I noticed a few little issues that would be worth polishing. They're not all necessarily blockers, so could potentially be things to look into in follow-ups if folks don't think they're deal breakers.

1. The Popover opens with a position that aligns the bottom of the popover with the bottom of the button. This looks good at first, but if you toggle between sizes, the height of the popover changes, which means that the position of the buttons on the screen changes, causing things to bump upwards. I'm wondering if we can (somehow) change where the position for the popover is locked to, so that if the height of the popover changes the vertical position of it doesn't? It could be a challenging balancing act, but ideally the buttons wouldn't move when toggling the controls. Here's a video of toggling between sizes:

https://github.com/WordPress/gutenberg/assets/14988353/98e9fdef-e9de-4437-8308-5a7fd90fe6d0

2. When dragging the focal point picker to the right edge of the popover, we get a horizontal scrollbar. This might not have been introduced in this PR, but I'm wondering if we should hide overflow-x for this popover (or something to the effect), to prevent this? It might only be visible on OSes that have set scrollbars to always display:

https://github.com/WordPress/gutenberg/assets/14988353/e46fd82b-abb7-4347-bbfa-3fd2870b7a06

3. On shorter screen heights, I found the popover to not position itself in a way that optimises its position on the screen. This could be a bit of a tricky one, but I think ideally on shorter screen heights, it'd fill as much of the viewport height as possible. There might still be a vertical scrollbar, though, so possibly something for us to watch out for with the height of this particular popover.

https://github.com/WordPress/gutenberg/assets/14988353/b7a70b57-f333-4df2-acf6-659134ff6f0a

4. On mobile screen sizes the popover is isn't visible (I suspect it's off the edge of the viewport)

| Color popover | Background image popover (I think it's off to the left) |
| --- | --- |
| <img width="328" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/bb4fed7f-058d-44a0-89c2-da05489f470f"> | <img width="321" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/52a86072-ac45-4002-ac85-4f03c5e9a01c"> |

Overall, though, I like the idea of the popover, it solves a lot of the issues with the very long controls we were encountering before!

ramonjd commented:
Thanks for testing @andrewserong 🙇🏻 and for the feedback

> The Popover opens with a position that aligns the bottom of the popover with the bottom of the button. This looks good at first, but if you toggle between sizes, the height of the popover changes, which means that the position of the buttons on the screen changes, causing things to bump upwards. I'm wondering if we can (somehow) change where the position for the popover is locked to, so that if the height of the popover changes the vertical position of it doesn't? It could be a challenging balancing act, but ideally the buttons wouldn't move when toggling the controls. 

You'll notice the color controls popover has does something similar when switching between gradient and color tabs (and if one of those tabs has more color palette options):

https://github.com/WordPress/gutenberg/assets/6458278/1586c0ae-ad18-4394-863a-9afab3392fe4

We might have some success with playing around with position or shift [popover props](https://github.com/WordPress/gutenberg/blob/55f3a98159a754cee5162a05852881fa168134a7/packages/components/src/popover/types.ts#L26-L26). 

I'll try it out.

My hunch is that it's inherent to the popover component. 

From what I can see the popover will do it best to position itself vertically, and scroll/not scroll according its contents and its own and the window height.

In relation to the popover contents, the focal point picker further complicates the matter - it will reflect the image's dimensions. Very tall images will result in very tall popovers.

We could mitigate these issues, by keeping the controls as static as possible. For example, the focal point picker image receives either get a max-height, or it is removed completely in favour of X/Y inputs. The latter isn't a popular opinion I understand.

I also had a version where the repeat and width controls were always visible, but disabled appropriately. That could reduce the bump.



>  On shorter screen heights, I found the popover to not position itself in a way that optimises its position on the screen. This could be a bit of a tricky one, but I think ideally on shorter screen heights, it'd fill as much of the viewport height as possible.

Similarly, this is the case for the color popover. As mentioned above, we might strike a balance with popover props. I'll give it the old college try. I'm not confident given how varied the content background controls are.

> This might not have been introduced in this PR, but I'm wondering if we should hide overflow-x for this popover (or something to the effect), to prevent this? It might only be visible on OSes that have set scrollbars to always display

Good one! I'll check if overflow tweaking helps 👍🏻 

> On mobile screen sizes the popover is isn't visible (I suspect it's off the edge of the viewport)

Oh strange, I thought I'd tested that and it was working. Something might have changed. I'll take another look.

andrewserong commented:
Thanks for taking a look! Feel free to skip any of those nits that I found if they don't feel too bad for now. Popovers are very tricky to get right and as you say, many of these points of friction might be in common for all our design tools popovers rather than this control in particular, so could always be worth looking at separately further down the track 🙂

ramonjd commented:
> On mobile screen sizes the popover is isn't visible (I suspect it's off the edge of the viewport)

I tried Firefox, Safari and Chrome and couldn't reproduce 🤔 I'll have another test next week when my brain is rested. Thanks!

<img width="354" alt="Screenshot 2024-06-28 at 4 26 11 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/cb957072-dfb1-4be0-8cab-dd95c45d6822">

andrewserong commented:
> I tried Firefox, Safari and Chrome and couldn't reproduce 🤔 I'll have another test next week when my brain is rested. Thanks!

After re-testing this morning, I can't reproduce this either! It's working nicely for me in both the post and site editors, at the block and global styles level:

| Block level | Global styles |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/cd933a94-d05f-446f-93be-85a86534d399) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/a7be22fc-8ea9-4e86-a8fb-8878b7dba637) |

Not sure what was going on when I was testing it last week. Maybe let's just keep an eye on it once this PR has landed?

ramonjd commented:
> > On mobile screen sizes the popover is isn't visible (I suspect it's off the edge of the viewport)
> 
> I tried Firefox, Safari and Chrome and couldn't reproduce 🤔 I'll have another test next week when my brain is rested. Thanks!

I tested this again on various viewport widths and browsers, and still couldn't reproduce. 

Nevertheless, I believe adding `expandOnMobile` makes for a better experience in mobile anyway — the tools are larger and more accessible — so I added that.

👍🏻 

https://github.com/WordPress/gutenberg/assets/6458278/415c4a13-7963-4a10-a111-69dcea4dc6cd

> On shorter screen heights, I found the popover to not position itself in a way that optimises its position on the screen. This could be a bit of a tricky one, but I think ideally on shorter screen heights, it'd fill as much of the viewport height as possible.

As mentioned, I'm not sure about this. The popover has some constraints. I've tried adding `resize: false` but that means users with short viewport heights won't be able to access the controls.

`shift: true` is already activated, and that's meant to optimize the position so that the popovers stay in view when meeting the viewport edges.

Ultimately, the height of the popup is down to the contents, in particular, the focal point picker. 

In my view, the options would be to reduce the height and, optionally, make the controls static, that is, disabled and no hidden/reveals. 

This could occur on desktop only since `expandOnMobile` provides a neater interface for mobile users.

For example, for desktop only:

- removing focal point picker image completely and just use position x/y input controls OR
- separating controls using tabs, e.g., Position, Size Options

It's tricky, but I guess it means weighing up how much value the popover provides verses making the controls stable and usable. 

Personally, I'm not 100% sold on the popover if it means we sacrifice usability to make it work.

ramonjd commented:
> After re-testing this morning, I can't reproduce this either! It's working nicely for me in both the post and site editors, at the block and global styles level:

Ah, I didn't see this comment before posting [above](https://github.com/WordPress/gutenberg/pull/60151#issuecomment-2198995244). 

Thanks for retesting!

What do you think about the `expandOnMobile` change?

andrewserong commented:
Thanks for the detailed digging in here!

> What do you think about the expandOnMobile change?

To me it feels like that makes the controls a bit large overall, as the position picker winds up taking extra space. This is more noticeable with an image that's vertically oriented. So I slightly lean toward the popover without `expandOnMobile`. Here's how it's looking with one of my test images: 

<img width="222" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/3884303d-40d2-4bf7-8d8b-60c88b2e4efa">

> In my view, the options would be to reduce the height and, optionally, make the controls static, that is, disabled and no hidden/reveals.

That could be worth trying. I like the idea of keeping the height of the popover stable while toggling.

> Personally, I'm not 100% sold on the popover if it means we sacrifice usability to make it work.

Agreed. Conceptually I really like the idea of the popover, but in practice, I'm finding it harder to use in manual testing in the current state. The main issues for me right now are:

* The popover moving position when toggling (sounds like this could be resolved by the idea of disabled fields)
* The height of the popover when using vertically oriented images

The latter point is more noticeable on a smaller screen. Here's a full screen window on a 13" laptop. In this case I always get a vertical scrollbar:

https://github.com/WordPress/gutenberg/assets/14988353/47b8bb92-f773-4e11-8dc4-09b4b1943254

Would it be worth adding a maximum height for the focal point picker at this stage? Or was the concern about what happens with very narrow images 🤔

> It's tricky, but I guess it means weighing up how much value the popover provides verses making the controls stable and usable.

Totally. I do feel like this PR is getting close, but those two main points of friction while testing on my laptop lean me toward `trunk` feeling a bit more stable in terms of usability in the current state. But if the focal point picker has a max height + the toggles etc are in a stable position? I'm curious if that'll swing things the other way 😄

Keen to hear what @jasmussen and @jameskoster think, too!

jameskoster commented:
Let's not lose sight of why we're trying the popover; the UI on trunk takes up way too much space, and perhaps more importantly; is not going to scale to support multiple background images.

So we either need to make the popover work, or try another pattern like a drilldown.

Could the popover 'hug' the contents when space allows? For instance in the screenshot below there's no reason to rely on scrolling:

<img width="300" alt="Screenshot 2024-07-01 at 10 04 12" src="https://github.com/WordPress/gutenberg/assets/846565/b082c805-9899-46e2-bcdf-0555b2b3e81d">

Adding a max-height to the focal point control, and disabling settings to stop jumping seem like good suggestions to try.

ramonjd commented:
> Adding a max-height to the focal point control, and disabling settings to stop jumping seem like good suggestions to try.

For sure, I'm going to take a stab at that. Thanks for chiming in @jameskoster!

> Let's not lose sight of why we're trying the popover; the UI on trunk takes up way too much space, and perhaps more importantly; is not going to scale to support multiple background images.

I agree that the controls monopolize the space in the sidebar, and that something needs to change. 

The only point I wanted to make is that we ought not to swap one problem (space) for another (UX headaches).  

There's always a compromise, so I'm confident we'll arrive at something that works for most folks. 👍🏻 

> Could the popover 'hug' the contents when space allows? For instance in the screenshot below there's no reason to rely on scrolling:

I'm not sure that's the scenario @andrewserong was reporting on. Rather, if I understood the feedback correctly, it was the janky behavior of the popover in shorter viewport heights. 

The popover has some sophisticated logic to position itself and adapt its height — so it does its best — but, if the contents of the popover are taller than the available window space, it loses the plot. 

As said, I'll try to make the internals of the popover more static. I think that would help a lot. 

Thanks!!

ramonjd committed:
Rolling back expanding popover in mobile.
Squishing all the controls, including giving the focal point picker a max-height

ramonjd commented:
The squishening!

| Desktop  | Mobile |
| ------------- | ------------- |
| <img width="300" alt="Screenshot 2024-07-02 at 1 26 16 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/b4c82365-19f7-4a8b-95df-55cc837111e6"> | <img width="300" alt="Screenshot 2024-07-02 at 1 27 35 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/7d4058de-2ac9-4917-bdf9-3658bf9748a2">  |

andrewserong commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -147,10 +147,12 @@ export function BackgroundImagePanel( {
 		return null;
 	}
 
+	/*
 	const defaultControls = getBlockSupport( name, [
 		BACKGROUND_SUPPORT_KEY,
 		'__experimentalDefaultControls',
 	] );
+*/
```
Tiny nit: looks like this can be removed now?

ramonjd replied:
Good spotting

andrewserong reviewed (approved):
Great work @ramonjd, and thanks for the guidance @jameskoster!

This is testing very nicely for me now:

✅ The position of the popover feels stable when toggling between sizes, thanks to the disabled fields
✅ The focal point position area no longer takes up very much room, making the controls feel more contained
✅ On smaller screen sizes (like my laptop screen) the popover feels reasonably sized and doesn't display a scrollbar

https://github.com/WordPress/gutenberg/assets/14988353/1c87e20b-3ac5-4ade-829a-aca6b138958f

Overall this very much crosses the threshold for me of feeling much nicer to use than what we have on `trunk`. In terms of the fixed height of the focal point picker I wonder if we could afford to give it a little more room potentially, but I reckon at this stage it'd be best to leave further tweaking to follow-ups since this is a pretty decent sized change in its current form.

Might be worth seeing what the designers think before landing, but overall this LGTM! Just left a tiny nit of a code comment that can be removed before merge.

🚀

ramonjd committed:
Remove commented-out code

jameskoster commented:
> I wonder if we could afford to give it a little more room potentially

I had the same thought. As a small tweak, could the height scale based on viewport height?

It would be really great if we could use the same unit control as you can see in the Padding setting here:

<img width="301" alt="Screenshot 2024-07-02 at 16 07 56" src="https://github.com/WordPress/gutenberg/assets/846565/e5846fa3-bfe4-4407-939f-0ea1a83c08d9">

The 'Size' segmented control should also use the new 40px sizing (`__next40pxDefaultSize`). 

With these changes I think we're good to go :)

ramonjd commented:
> The 'Size' segmented control should also use the new 40px sizing (__next40pxDefaultSize).

I'll revert back, but note that removing `__next40pxDefaultSize` provides much of the space saving. 


> I wonder if we could afford to give it a little more room potentially
> 
> > I had the same thought. As a small tweak, could the height scale based on viewport height?

I did try `vh` with various image sizes, but it's a bit hit and miss. I'll try to land on a value. We can always change it later.


Regardless, I think we're in a good place to iterate. 

I'll fix the above issues and merge.


Thanks again for all the help!

ramonjd committed:
Revert input sizes
Ensure long file names are not cut off by min-width
Use clamp value for focal picker image height

ramonjd commented:
Okay! I think I've come up with a good compromise by using a clamp value for the focal picker image height. It can be tweaked later if needed.

<img width="312" alt="Screenshot 2024-07-03 at 10 04 43 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/9c4ae7b6-1bb4-4a5f-ab88-3c2ca8adbb81">
<img width="297" alt="Screenshot 2024-07-03 at 10 04 55 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/f9ebf439-2604-4367-a00c-9a8fb08b8912">
<img width="297" alt="Screenshot 2024-07-03 at 10 05 03 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/0458ba60-bb8f-4347-bda7-3a356948deca">

andrewserong commented:
> I think I've come up with a good compromise by using a clamp value for the focal picker image height.

Nice one! 🚀

jasmussen commented:
> I'll revert back, but note that removing __next40pxDefaultSize provides much of the space saving.

Just wanted to comment on this, the default size is 36px, and should ideally be avoided (see #46734) in favor of a taller and more legible 40px default, or exactly for space-saving concerns the more `compact` 32px.
